src/Domain/PaymentUrlGenerator/CreditCard.php

```diff
@@ -4,33 +4,39 @@
 
 namespace WMDE\Fundraising\PaymentContext\Domain\PaymentUrlGenerator;
 
+use WMDE\Fundraising\PaymentContext\Domain\Model\CreditCardPayment;
+use WMDE\Fundraising\PaymentContext\Domain\Model\Payment;
+
 /**
  * @license GPL-2.0-or-later
  * @author Kai Nissen < kai.nissen@wikimedia.de >
  */
 class CreditCard implements PaymentProviderURLGenerator {
 
 	private CreditCardConfig $config;
-	private AdditionalPaymentData $additionalPaymentData;
+	private CreditCardPayment $payment;
 
-	public function __construct( CreditCardConfig $config, AdditionalPaymentData $additionalPaymentData ) {
+	public function __construct( CreditCardConfig $config, Payment|CreditCardPayment $payment ) {
+		if ( !$payment instanceof CreditCardPayment ) {
+			throw new \LogicException( "Must be of type CreditCardPayment, you provided: " . get_class( $payment ) );
```

Comment:
Wow, this must have been really frustrating bashing the code until both PHP and PHPStan are happy. I appreciate your effort and brainpower, trying to be generic enough for the factory to work and specific enough for the generator code to work. I'm worried though, that our future selves might be confused (and maybe appalled).

The runtime type check and union types in the generator constructors are unnecessary for PHP to work. PHP would work fine specifying the child classes in the constructor, because our factory code already does the type check before it passes on the instance. It's just PHPStan that doesn't recognize the type checks yet when we do them with `get_class` instead of `instanceof`. I looked into this, wrote example code and finally found this bug from 2017: https://github.com/phpstan/phpstan/issues/523

So the question boils down to "which ugly construct do we use to appease our static type checker, which one is more readable and has less performance impact?" In this case I'd go with a child class in the URL generator constructors (making the constructor shorter and less ambiguous, removing the need for type checks -> better performance) and code with the following pattern in the factory:

```php
if ($payment instanceof YYYPayment) {
     return new YYYGenerator($payment);
} elseif ($payment instanceof ZZZPayment) {
    return new ZZZGenerator($payment);
} 
return NullGenerator();
``` 
With just 4 branches its manageable enough.